src/client/internal.ts

```diff
@@ -20,29 +24,17 @@ export const SERVICE_CHANNEL = 'of-fdc3-service-v1';
 export enum APITopic {
     OPEN = 'OPEN',
     FIND_INTENT = 'FIND-INTENT',
-    FIND_INTENTS_BY_CONTEXT = 'FIND-INTENT-BY-CONTEXT',
+    FIND_INTENTS_BY_CONTEXT = 'FIND-INTENTS-BY-CONTEXT',
     BROADCAST = 'BROADCAST',
     RAISE_INTENT = 'RAISE-INTENT'
 }
 
-export interface TopicPayloadMap {
-    [APITopic.OPEN]: OpenPayload;
-    [APITopic.FIND_INTENT]: FindIntentPayload;
-    [APITopic.FIND_INTENTS_BY_CONTEXT]: FindIntentsByContextPayload;
-    [APITopic.BROADCAST]: BroadcastPayload;
-    [APITopic.RAISE_INTENT]: RaiseIntentPayload;
-}
-
-export interface TopicResponseMap {
-    [APITopic.OPEN]: void;
-    [APITopic.FIND_INTENT]: AppIntent;
-    [APITopic.FIND_INTENTS_BY_CONTEXT]: AppIntent[];
-    [APITopic.BROADCAST]: void;
-    // This is not strictly to spec as target should be able to return data.
-    // Needs a not-insignificant amount of provider changes to allow the two-way
-    // transmission of data before it can be put back to IntentResponse
-    // TODO: Revisit after SERVICE-392
-    [APITopic.RAISE_INTENT]: void;
+export type API = {
+    [APITopic.OPEN]: [OpenPayload, void];
```

Comment:
Using `{payload, response}` would work too, I figured the tuple was a bit more concise - makes the actual API a bit easier to see in a 
`[FindIntentPayload, AppIntent]` => `(payload: FindIntentPayload) => AppIntent` kinda way.

Open to either - for reference, using the latter API definition would look like: 

```
interface API extends APISpecification<APITopic> {
    [APITopic.OPEN]: {payload: OpenPayload, response: void};
    [APITopic.FIND_INTENT]: {payload: FindIntentPayload, response: AppIntent};
    [APITopic.FIND_INTENTS_BY_CONTEXT]: {payload: FindIntentsByContextPayload, response: AppIntent[]};
    [APITopic.BROADCAST]: {payload: BroadcastPayload, response: void};
    [APITopic.RAISE_INTENT]: {payload: RaiseIntentPayload, response: IntentResolution};
}
```

As for `APISpecification`, seems we can add an "extends" (as above) without breaking any of the type checking. Seems like TypeScript doesn't really enforce that `API` has all the right keys, but the "integrity" of `API` still gets enforced when it's used (such as when calling `registerListeners`). Even if the type checking isn't perfect, having the "hint" there is definitely a bonus.

Reply:
Heh, managed to confuse myself a bit here - `API` was actually being defined twice...

Once in `internal.ts`, where it's defined as a `type` and so can't use `APISpecification` (since `type` declarations have no kind of "extends" mechanism), and again in `index.ts` where it's defined as an `interface` (which means it can extend `APISpecification`).

Have removed the duplicate definitions in index.ts - but now, as you point out, can't really reference `APISpecification` here. It's possible if we move the declarations of all those `API*` types from the provider side to the client side, but would prefer to keep those out of the client if possible. Would also make breaking-out `APIHandler` into a separate module (as it's a good candidate for something that should be standardised across services) more complicated.

Will defer this one for now.

Reply:
In my view this comes down to a balance of two competing interests: magic indices make definitions more straightforward but referencing more confusing, whereas object-style makes definitions uglier while making referencing clearer and less error prone. 

I guess the correct solution would depend on whether we expect to do more defining or more referencing over the whole project.

Reply:
True. I'd prefer cleaner definitions, particularly when we start making breaking changes that we have to maintain backward compatibility for. When we get to that point, we're going to have multiple API definitions (v1, v2, etc) for each service.

APIHandler is still pretty simple, some extra documentation in there that allows for more readable definitions within the "service" itself would be my preference.